File: frontend/src/api/gitHistory.ts
export interface CommitListItem {
  id: number;
  commit_hash: string;
  commit_date: string;
  author: string;
  message: string;
  repository_id: number;
  repository_name: string;
  changed_file_count: number;
  additions: number;
  deletions: number;
}

export interface CommitListResponse {
  items: CommitListItem[];
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
}

export interface CommitChange {
  id: number;
  file_path: string;
  change_type: string;
  additions: number | null;
  deletions: number | null;
  diff: string | null;
}

export interface CommitDetail {
  id: number;
  equipment_id: number;
  commit_hash: string;
  commit_date: string;
  author: string;
  message: string;
  parent_hash: string | null;
  changes: CommitChange[];
}

export interface CommitSearchParams {
  q?: string;
  date_from?: string;
  date_to?: string;
  file_path?: string;
  author?: string;
  repository_id?: number;
  page?: number;
  page_size?: number;
}

export interface GitSyncResult {
  equipment_id: number;
  scanned_commits: number;
  new_commits: number;
  skipped_commits: number;
  new_changes: number;
  status: string;
}

const API_BASE = import.meta.env.VITE_API_BASE ?? "";

async function parseError(response: Response): Promise<string> {
  try {
    const data = await response.json();
    if (typeof data.detail === "string") return data.detail;
  } catch {
    /* ignore */
  }
  return `요청 실패 (${response.status})`;
}

function buildQuery(params: CommitSearchParams): string {
  const qs = new URLSearchParams();
  if (params.q) qs.set("q", params.q);
  if (params.date_from) qs.set("date_from", params.date_from);
  if (params.date_to) qs.set("date_to", params.date_to);
  if (params.file_path) qs.set("file_path", params.file_path);
  if (params.author) qs.set("author", params.author);
  if (params.repository_id) qs.set("repository_id", String(params.repository_id));
  if (params.page) qs.set("page", String(params.page));
  if (params.page_size) qs.set("page_size", String(params.page_size));
  const s = qs.toString();
  return s ? `?${s}` : "";
}

export async function fetchCommitList(
  equipmentId: number,
  params: CommitSearchParams = {},
): Promise<CommitListResponse> {
  const response = await fetch(
    `${API_BASE}/api/equipment/${equipmentId}/git/commits${buildQuery(params)}`,
  );
  if (!response.ok) throw new Error(await parseError(response));
  return response.json();
}

export async function fetchCommitDetail(commitId: number): Promise<CommitDetail> {
  const response = await fetch(`${API_BASE}/api/git/commits/${commitId}`);
  if (!response.ok) throw new Error(await parseError(response));
  return response.json();
}

export async function syncGitHistory(equipmentId: number): Promise<GitSyncResult> {
  const response = await fetch(`${API_BASE}/api/equipment/${equipmentId}/sync/git`, {
    method: "POST",
  });
  if (!response.ok) throw new Error(await parseError(response));
  return response.json();
}

export { fetchEquipmentList, type Equipment } from "./equipment";

File: backend/app/schemas/git_history.py
import math
from dataclasses import dataclass

from pydantic import BaseModel, Field


class GitSyncResponse(BaseModel):
    equipment_id: int
    scanned_commits: int
    new_commits: int
    skipped_commits: int
    new_changes: int
    status: str


class CommitListItem(BaseModel):
    id: int
    commit_hash: str
    commit_date: str
    author: str
    message: str
    repository_id: int
    repository_name: str
    changed_file_count: int
    additions: int
    deletions: int


class CommitListResponse(BaseModel):
    items: list[CommitListItem]
    page: int
    page_size: int
    total: int
    total_pages: int


class CommitChangeDetail(BaseModel):
    id: int
    file_path: str
    change_type: str
    additions: int | None
    deletions: int | None
    diff: str | None


class CommitDetailResponse(BaseModel):
    id: int
    equipment_id: int
    repository_id: int
    repository_name: str
    commit_hash: str
    commit_date: str
    author: str
    message: str
    parent_hash: str | None
    changes: list[CommitChangeDetail]


@dataclass
class CommitSearchParams:
    equipment_id: int
    repository_id: int | None = None
    q: str | None = None
    date_from: str | None = None
    date_to: str | None = None
    file_path: str | None = None
    author: str | None = None
    page: int = 1
    page_size: int = 50


MAX_PAGE_SIZE = 200
DEFAULT_PAGE_SIZE = 50


def normalize_page_size(page_size: int) -> int:
    if page_size < 1:
        return DEFAULT_PAGE_SIZE
    return min(page_size, MAX_PAGE_SIZE)


def normalize_page(page: int) -> int:
    return max(page, 1)


def calc_total_pages(total: int, page_size: int) -> int:
    if total == 0:
        return 0
    return math.ceil(total / page_size)
